File: go/crypto-square/crypto_square.go
package cryptosquare

import (
	"math"
	"strings"
	"unicode"
)

func Encode(pt string) string {
	// Handle empty input
	if len(pt) == 0 {
		return ""
	}

	// Normalize the input: remove spaces and punctuation, convert to lowercase
	normalized := normalize(pt)
	if len(normalized) == 0 {
		return ""
	}

	// Calculate the dimensions of the rectangle
	c, r := calculateDimensions(len(normalized))

	// If we only have one character, return it as is
	if len(normalized) == 1 {
		return normalized
	}

	// Create the rectangle
	rectangle := createRectangle(normalized, r, c)

	// Read down the columns to create the encoded message
	return encodeRectangle(rectangle, r, c)
}

// normalize removes spaces and punctuation, and converts to lowercase
func normalize(s string) string {
	var result strings.Builder
	for _, ch := range s {
		if unicode.IsLetter(ch) || unicode.IsDigit(ch) {
			result.WriteRune(unicode.ToLower(ch))
		}
	}
	return result.String()
}

// calculateDimensions calculates the dimensions of the rectangle
func calculateDimensions(length int) (c, r int) {
	// Find the smallest possible integer c such that:
	// - r * c >= length
	// - c >= r
	// - c - r <= 1
	c = int(math.Ceil(math.Sqrt(float64(length))))
	r = int(math.Ceil(float64(length) / float64(c)))
	return c, r
}

// createRectangle arranges the normalized text into a rectangle
func createRectangle(text string, rows, cols int) [][]rune {
	rectangle := make([][]rune, rows)
	for i := range rectangle {
		rectangle[i] = make([]rune, cols)
		// Initialize with spaces
		for j := range rectangle[i] {
			rectangle[i][j] = ' '
		}
	}

	// Fill the rectangle with the text
	for i, ch := range text {
		row := i / cols
		col := i % cols
		if row < rows && col < cols {
			rectangle[row][col] = ch
		}
	}

	return rectangle
}

// encodeRectangle reads down the columns to create the encoded message
func encodeRectangle(rectangle [][]rune, rows, cols int) string {
	var result strings.Builder
	
	for col := 0; col < cols; col++ {
		if col > 0 {
			result.WriteRune(' ')
		}
		
		for row := 0; row < rows; row++ {
			if row < len(rectangle) && col < len(rectangle[row]) {
				result.WriteRune(rectangle[row][col])
			}
		}
	}
	
	return result.String()
}

File: cpp/crypto-square/crypto_square.cpp
#include "crypto_square.h"
#include <algorithm>
#include <cctype>
#include <cmath>
#include <sstream>

namespace crypto_square {

cipher::cipher(const std::string& text) : plaintext_(normalize(text)) {}

std::string cipher::normalize(const std::string& text) const {
    std::string normalized;
    for (char c : text) {
        if (std::isalnum(c)) {
            normalized.push_back(std::tolower(c));
        }
    }
    return normalized;
}

std::string cipher::normalized_cipher_text() const {
    if (plaintext_.empty()) {
        return "";
    }

    // Calculate the dimensions of the rectangle
    int length = plaintext_.length();
    int c = std::ceil(std::sqrt(length)); // number of columns
    int r = std::ceil(static_cast<double>(length) / c); // number of rows

    // Ensure c >= r and c - r <= 1
    if (c < r) {
        c = r;
    }

    // Create the encoded text by reading down the columns
    std::stringstream result;
    
    // For each column position
    for (int col = 0; col < c; ++col) {
        // For each row in this column
        for (int row = 0; row < r; ++row) {
            int index = row * c + col;
            if (index < length) {
                result << plaintext_[index];
            } else {
                // Add a space if we're in the padding area
                result << " ";
            }
        }
        
        // Add a space between chunks, but not after the last chunk
        if (col < c - 1) {
            result << " ";
        }
    }

    return result.str();
}

}  // namespace crypto_square
